### PR TITLE
[Merged by Bors] - refactor(RepresentationTheory/GroupCohomology): tidy up 8599 and add more low degree cocycle API

### PR DESCRIPTION
**Pull request by 101damnations**
---

Sorry to have merged #8599 hastily! This PR tidies up details like namespacing in that PR. Also, given an additive or multiplicative group `A` with an appropriate scalar action of a group `G`, we define low degree cocycle/coboundary conditions on $Fun(G, A), Fun(G^2, A),$ and then link that up with the existing cocycle/coboundary types for the representation associated to `A`. Using this, `RepresentationTheory.GroupCohomology.Hilbert90` gets clearer.

Lots of `DistribMulAction` and `MulDistribMulAction` stuff here but it seems like they can't be related by `to_additive` afaik.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alexjbest** [comment]
The module docstring of https://leanprover-community.github.io/mathlib4_docs/Mathlib/RepresentationTheory/GroupCohomology/Basic.html#groupCohomology has some broken tex in it, could you also fix that in this PR? or later

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean
Comment (erdOne) [inline Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean:78]: I think we usually give it a name like `isMulOneCoboundary_of_isMulOneCocycle_of_aut_to_units`

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (erdOne) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:401]: Can you restate `mem_oneCocycles_iff` etc. as an iff using the predicates introduced above?

Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:454]: Thanks for the comments! I have done this now

**Comment by alexjbest** [comment]
LGTM, sorry that this was waiting so long now!

maintainer merge

**Comment by alexjbest** [comment]
Oh scrap that we have a merge conflict I guess, but after that is fixed let's merge this

**Comment by 101damnations** [comment]
Thank you! I have fixed the conflict

**Comment by alexjbest** [comment]
maintainer merge

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean
Comment (jcommelin) [inline Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean:27]: It looks like all mentions of Hilbert90 have left this file... please mention it in comments/docstrings.

Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean]: done, thanks!

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean:85]: I think people use nowadays `<|` more than `$`, but as you prefer.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:324]:
```suggestion
def oneCoboundariesOfMemRange {f : G → A} (h : f ∈ LinearMap.range (dZero A)) :
```
Not very important.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:329]: Do you mean that `f` should be implict and `h` explicit? The other way seems very weird.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:334]: `f` can be implicit, but as you wish.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:351]: Again `f` can be implicit and similarly below.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:356]: As above it seems weird. This happens also below.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:338]: Same.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:383]: Is this lemma autogenerated by `@[mk_iff]`?

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:391]: Same.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:402]:
```suggestion
  simpa only [mul_one, one_smul, self_eq_add_right] using (isOneCocycle_iff f).1 hf 1 1
```
untested. If it works you can also remove the `by`. Same below.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:433]:
```suggestion
end IsCocycle

section IsCoboundary

variable {G A : Type*} [Mul G] [AddCommGroup A] [SMul G A]
```
I find it more readable, but do as you prefer.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:503]:
```suggestion
end ofDistribMulAction

section IsMulCocycle

section

variable {G M : Type*} [Mul G] [CommGroup M] [SMul G M]
```
This section is the same as above but for multiplicative groups, right? If this is the case we probably want a docstring explaining it.

**Review by riccardobrasca — APPROVED** [review]

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:329]: I was under the impression that when a hypothesis is used to define something in the lemma statement, it can be left implicit

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:334]: I know `f`'s in the type of `hf`, but I feel like I've been annoyed in the past by things with implicit data arguments where I then have to start the proof of some hypothesis with `show...`. I can't remember the context though so it may be irrelevant, and I guess in Lean 4 it's easier to just specify `f` with `(f := ...)` if you want to.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:383]: Only applies to inductive declarations, apparently. Maybe I should make `IsOneCocycle` etc a class?

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:402]: This works, thank you. Confused about the `by` comment, though

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (alreadydone) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:383]: I think you can simply use `rw [IsOneCocycle]` and remove this lemma.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (alreadydone) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:402]: Does `simpa only [mul_one, one_smul, self_eq_add_right] using hf 1 1` work?

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (alreadydone) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:329]: `f` appears in both sides of equality so I'd make it implicit. `h` doesn't appear in RHS so I'd make it explicit, for ease of rewriting from right to left.

Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:329]: I see, thank you!

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:383]: Fair play, will do

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean
Comment (riccardobrasca) [reply Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean]: If you find this version more convenient just keep it like that.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**
